### ISSUE TITLE
Traversable.mapAccumLeft

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#4443]
Question 1:

Is this code equivalent to `Traversable.mapAccumLeft`?:

```
pub def mapAccumLeft2(f: (acc, a) -> (acc, b) & ef, start: acc, t: t[a]): (acc, t[b]) & ef with Traversable[t] =
    region r {
        let acc = ref start @ r;
        let tb = t |> Functor.map(x -> {
            let (nacc, b) = f(deref acc, x);
            acc := nacc;
            b
        });
        (deref acc, tb)
    }
```

Question 2:

Does this mean `mapAccumLeft` only needs `Functor` and should be moved to a different namespace?

**Comment by magnus-madsen** [comment]
@stephentetley Thoughts?

@JonathanStarup Can we use this as an example for the region paper? Can you try to look at `mapAccumRight` and see if it can be expressed in a similar way?

**Comment by stephentetley** [comment]
Question 1

Yes (I think) if the functor implementation of `map` goes left to right (caveat below).

Question 2

Probably not. The documentation for Haskell's `Traversable` class makes the strong requirement that the traversal should be left-to-right, I missed this off when I implemented the Flix version and should have included it.  

`mapAccum` effectively carries a trace of previous states which can be a summary of nodes visited, so it is essential to the idea of the function that the visit order is defined. `Functor.map` 's traversal order is unspecified so we can't safely rely on it for a `mapAccumLeft` or `mapAccumRight`.

**Comment by magnus-madsen** [comment]
Hmm. What about using Foldable? There we have fold left and right

**Comment by stephentetley** [comment]
> Foldable

I don't think so - you want both the container (rebuilt / mapped) and the summary value. You can't rebuild the container with just `Foldable` the fold step would need to know about a `cons` operation.

**Comment by magnus-madsen** [comment]
Oh right

**Comment by magnus-madsen** [comment]
What happens if I traverse with `Identity`? Would that work? But then I guess I would not know how to accum from the right. 

Do you use these two functions?

(I bring these things up because the implementation is causing us a few problems with the effect system).

**Comment by stephentetley** [comment]
I've made a version of `Traversable` where `ContStateLeft` and `ContStateRight` thread a region rather than IO so the effects of `map` and `ap` can be internally cast to the region but pushed up again to the outer-level so they aren't lost (this is the same trick to track the effects as before but with a region rather than IO/Impure). 

I don't know if this will be more benign to the type system. I'll make a PR so I don't lose the code, but I'm happy if you don't think its worthwhile (neither version seems ideal). 

If the PR isn't an improvement I can live without general versions of `mapAccumLeft` and `mapAccumRight` in Traversable for the time being until there is a safe way for "function monads" to live with effects.

**Comment by stephentetley** [comment]
I've made the PR - #4460.

**Comment by magnus-madsen** [comment]
What about this?:

```
pub def mapAccumLeft2(f: (acc, a) -> (acc, b) & ef, start: acc, t: t[a]): (acc, t[b]) & ef with Traversable[t] =
    region r {
        let acc = ref start @ r;
        let Identity(tb) = t |> Traversable.traverse(x -> {
            let (nacc, b) = f(deref acc, x);
            acc := nacc;
            Identity(b)
        });
        (deref acc, tb)
    }
```


But I don't see how to get accumRight from this. Why is it even possible to express accumRight with Traversable? Is it because you can use some kind of continuation monad that satisfies the right laws and  then use it to build the reverse traversal?

**Comment by stephentetley** [comment]
I don't really understand why StateRight works - you can implement it without CPS but then it isn't stack safe. Basically it has the evaluation order of `Applicative.ap` changed so the applicative "value" is run before the applicative "function" - this seems to be what's require to go rightwards.

**Comment by stephentetley** [comment]
> But I don't see how to get accumRight from this.

We could just provide `mapAccumLeft` as Traversable should be explicitly left-to-right. With a ref cell for the interim state we would need a `traverseRight` method to get `mapAccumRight`.

**Comment by magnus-madsen** [comment]
Can we live without mapAccumRight?

**Comment by stephentetley** [comment]
> Can we live without mapAccumRight?

I'm happy without it, until there's a good solution for the "function monads" and effects problem...
